### ISSUE TITLE
Suppor for client build with "client-secret"

### ISSUE THREAD

**Issue by napayaofenkai** [okta/okta-sdk-java#424]
After reading your document, I can't understand how to build a client with oauth2, It confuses me that "pathToYourPrivateKeyPemFile", becaust in my YMAL, I use like this
```
okta:
  oauth2:
    issuer: https://dev-524104.okta.com/oauth2/default
    client-id: xxxxxxxxxxxxx
    client-secret: xxxxxxxxxxxxxxxxxxxx
    redirect-uri: /authorization-code/callback
```
Why not client build with client-secret but a PemFile? And you don't tell my how to get and set the PemFile

Your Document：
```
Client client = Clients.builder()
    .setOrgUrl("{yourOktaDomain}")
    .setAuthorizationMode(AuthorizationMode.PRIVATE_KEY)
    .setClientId("{clientId}")
    .setScopes(new HashSet<>(Arrays.asList("okta.users.read", "okta.apps.read")))
    .setPrivateKey("{pathToYourPrivateKeyPemFile}")
    .build();
```

**Comment by arvindkrishnakumar-okta** [comment]
@napayaofenkai Thanks for your question!

private_key_jwt client authentication method is the only supported method for clients that want to get access tokens with Okta scopes. Please refer to https://developer.okta.com/docs/guides/implement-oauth-for-okta-serviceapp/create-publicprivate-keypair/ for help on how to create a private key file in PEM format.

Once you have created the private key pem file, you could supply its absolute file path either in YAML config or set it programmatically in the client builder. 

Sample YAML config:

`
okta:
  client:
    connectionTimeout: 30 # seconds
    orgUrl: "https://{yourOktaDomain}" # i.e. https://dev-123456.oktapreview.com
    proxy:
      port: null
      host: null
      username: null
      password: null
    authorizationMode: "PrivateKey"
    clientId: "yourClientId"
    scopes: "okta.users.read okta.apps.read"
    privateKey: "/path/to/yourPrivateKey.pem" # PEM format. This SDK supports RSA AND EC algorithms - RS256, RS384, RS512, ES256, ES384, ES512.
    requestTimeout: 0 # seconds
    rateLimit:
      maxRetries: 4
`

If you like to use the client builder and set it programmatically, the code would look like:

`
Client client = Clients.builder()
    .setOrgUrl("https://{yourOktaDomain}")  // e.g https://dev-123456.okta.com
    .setAuthorizationMode(AuthorizationMode.PRIVATE_KEY)
    .setClientId("yourClientID")
    .setScopes(new HashSet<>(Arrays.asList("okta.users.read", "okta.apps.read")))
    .setPrivateKey("/path/to/yourPrivateKey.pem")
    .build();
`

You might also want to refer https://github.com/okta/okta-sdk-java#configuration-reference for notes on config precedence.

Let me know if you have further questions on this subject.

**Comment by napayaofenkai** [comment]
Thank you for your guidance, but I think it inconvenient to use in this way. Since I have `client_id` and `client-secret`, Is it better if I can use these two parameters to successfully create the client? Like fellow:  
```
okta:
  oauth2:
    issuer: https://dev-524104.okta.com/oauth2/default
    client-id: client_id
    client-secret: agadknfaasd_SSs
    redirect-uri: /authorization-code/callback

Client client = Clients.builder()
    .setOrgUrl("https://{yourOktaDomain}")  // e.g https://dev-123456.okta.com
    .setAuthorizationMode(AuthorizationMode.PRIVATE_KEY)
    .setClientId({client-id})
    .setScopes(new HashSet<>(Arrays.asList("okta.users.read", "okta.apps.read")))
    .setPrivateKey("agadknfaasd_SSs")
    .build();
```

**Comment by bdemers** [comment]
Hi @napayaofenkai!
Thanks for reaching out!

It looks like you are also using our Spring Boot integration?  If that is the case reusing the user's existing token _might_ be your best bet: #339 and my comment in: https://github.com/okta/okta-sdk-java/pull/339#issuecomment-654277671

This depends on the type of client you are building and if the requests made to Okta are on-behalf-of a user (for example you need to get the list of applications that the current user has access to).  Otherwise, if you don't need to scope the request to the current user, then the SDK might have a different client-id and secret from your Spring Boot application.

TL;DR - We would love to hear more about your use-case, please add a comment to [this thread](https://github.com/okta/okta-sdk-java/pull/339#issuecomment-654277671)

**Comment by napayaofenkai** [comment]
> Hi @napayaofenkai!
> Thanks for reaching out!
> 
> It looks like you are also using our Spring Boot integration? If that is the case reusing the user's existing token _might_ be your best bet: #339 and my comment in: [#339 (comment)](https://github.com/okta/okta-sdk-java/pull/339#issuecomment-654277671)
> 
> This depends on the type of client you are building and if the requests made to Okta are on-behalf-of a user (for example you need to get the list of applications that the current user has access to). Otherwise, if you don't need to scope the request to the current user, then the SDK might have a different client-id and secret from your Spring Boot application.
> 
> TL;DR - We would love to hear more about your use-case, please add a comment to [this thread](https://github.com/okta/okta-sdk-java/pull/339#issuecomment-654277671)!

  Hi, @bdemers , thank you for your help. In fact, I know that I can use accessToken to access client.  
But in my situtation, I'm not the admin of okta, and accessToken has an expiration date. If I use accessToken to build my client, that means I have to modify my code or cofig propreties frequently. And it's not easy for me to get a new token because I'm not the okta admin.

**Comment by bdemers** [comment]
@napayaofenkai 

I'm not sure I'm following, it sounds like you might be mixing short-lived credentials (accessToken) with long-lived credential (client-secret).

You wouldn't configure your application with an accessToken, you would either let the SDK manage it for you, or as proposed in #339, use an access token retrieved from a different source, i.e. the currently logged in user.

Does that help?

**Comment by arvindkrishnakumar-okta** [comment]
Closing this due to inactivity, feel free to reopen if needed.
